[WebGPU] Implement deferred compilation for RenderPipeline

tadeuzagallo opened the pull request:
#### aefb962a382399124429bcfaf40325debec4848c
<pre>
[WebGPU] Implement deferred compilation for RenderPipeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=253520">https://bugs.webkit.org/show_bug.cgi?id=253520</a>
rdar://106365226

Reviewed by Mike Wyrzykowski and Myles C. Maxfield.

So far, for render pipelines, we only support compiling if layout hints are
provided at ShaderModule-creation time. This implements deferred compilation
when creating the pipeline if no hints were provided.

* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
(WebGPU::createLibrary): Deleted.
(WebGPU::createConstantValues): Deleted.
(WebGPU::createFunction): Deleted.
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):

Canonical link: <a href="https://commits.webkit.org/261874@main">https://commits.webkit.org/261874@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d160ae20a361aabf18b2be3953ff1d59b42fa53

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113122 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22254 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/90/builds/1792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4893 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121585 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/23625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/13437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/6138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118909 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/23625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/90/builds/1792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106204 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/23625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/90/builds/1792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14557 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/90/builds/1792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/98833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15272 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/13437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20565 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/90/builds/1792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8291 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17118 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg reviewed (approved):
:shipit:

I'd prefer to see refactoring such as creation of `PipelineInlines.h` as separate commits so I don't have to parse out what is functional change from just moving code around, but I understand, this is not a very WebKit-y thing. (I'd be happy to review multiple commits that are squashed on landing as a compromise)

litherum commented on Source/WebGPU/WebGPU/PipelineInlines.h:
```diff
@@ -0,0 +1,123 @@
+    WGSL::Reflection::EntryPointInformation entryPointInformation; // FIXME(PERFORMANCE): This is big. Don't copy this around.
+};
+
+static std::optional<LibraryCreationResult> createLibrary(id<MTLDevice> device, const ShaderModule& shaderModule, const PipelineLayout* pipelineLayout, const String& entryPoint, NSString *label)
```
A few thoughts:

1. The name of this file is called "...Inlines.h" but these functions are not `inline`
2. I don't think it's right for `static` non-`inline`d functions to be duplicated in multiple compilation units
3. Do these functions even need to be inlined anyway? Can we just use the normal `.h`/`.cpp` pair for them?

mwyrzykowski replied:
I would prefer (3) unless there is a reason not to, though I don't think there is a problem with multiple compilation units having the same definition of a function other than it wastes memory. `inline` here would make the problem worse in terms of memory.

mwyrzykowski replied:
assuming `inline` is honored and so on otherwise it becomes the same as it appears now

tadeuzagallo committed:
[WebGPU] Implement deferred compilation for RenderPipeline
https://bugs.webkit.org/show_bug.cgi?id=253520
rdar://106365226

Reviewed by Mike Wyrzykowski and Myles C. Maxfield.

So far, for render pipelines, we only support compiling if layout hints are
provided at ShaderModule-creation time. This implements deferred compilation
when creating the pipeline if no hints were provided.

* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
(WebGPU::createLibrary): Deleted.
(WebGPU::createConstantValues): Deleted.
(WebGPU::createFunction): Deleted.
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):

Canonical link: https://commits.webkit.org/261874@main
